File: Architecture/TodoApp/android/event/src/main/java/com/novoda/event/EventObserver.java
package com.novoda.event;

import rx.Observer;

public abstract class EventObserver<T> implements Observer<Event<T>> {

    @Override
    public void onNext(Event<T> event) {
        switch (event.state()) {
            case IDLE:
                onIdle(event);
                break;
            case LOADING:
                onLoading(event);
                break;
            case ERROR:
                onError(event);
                break;
            default:
                break;
        }
    }

    @Override
    public void onError(Throwable e) {
        throw new IllegalStateException("Event Pipeline failed, this should never happen", e);
    }

    @Override
    public void onCompleted() {
        throw new IllegalStateException("Event Pipeline completed, this should never happen");
    }

    public abstract void onLoading(Event<T> event);

    public abstract void onIdle(Event<T> event);

    public abstract void onError(Event<T> event);

}

File: Architecture/UnidirectionalKotlinCats/core/src/main/kotlin/com/odai/architecturedemo/event/EventObserver.kt
package com.odai.architecturedemo.event

import rx.Observer

abstract class EventObserver<T>: Observer<Event<T>> {

    override fun onError(p0: Throwable?) {
        throw UnsupportedOperationException("Error on event pipeline. This should never happen", p0)
    }

    override fun onCompleted() {
        throw UnsupportedOperationException("Completion on event pipeline. This should never happen")
    }

    override fun onNext(p0: Event<T>) {
        when (p0.status) {
            Status.LOADING -> onLoading(p0)
            Status.IDLE -> onIdle(p0)
            Status.ERROR -> onError(p0)
        }
    }

    abstract fun onLoading(event: Event<T>);

    abstract fun onIdle(event: Event<T>);

    abstract fun onError(event: Event<T>);
}
